File: CC_01 Sandpile/JS/Sandpile.js
function Sandpile(values) {

	this.matrix = values;
	this.size = this.matrix.length;
	this.maxValue = 3;
	this.squareSize = 500/this.size;
	this.colors = ['rgb(255, 0, 0)', 'rgb(0,255,0)', 'rgb(0, 0, 255)', 'rgb(255, 255, 0)'];

	this.show = function() {
		var margin = 50;
		console.log(margin);
		var rectSize = (width-2*margin) / this.size;

		for (i = 0; i < this.size; i++) {
			for (j = 0; j < this.size; j++) {
				fill(this.colors[this.matrix[i][j]]);
				rect(j*rectSize+margin, i*rectSize+margin, rectSize, rectSize);
				fill(0);
				text(this.matrix[i][j], j*rectSize+margin, i*rectSize+margin, rectSize, rectSize);
			}
		}
	}


	this.stable = function() {
		for (i = 0; i < this.size; i++) {
			for (j = 0; j < this.size; j++) {
				if (this.matrix[i][j] > this.maxValue) {
					return false;
				}
			}
		}
		return true;
	}


	this.topple = function(values) {
		while (!this.stable()) {
			for (i = 0; i < this.size; i++) {
				for (j = 0; j < this.size; j++) {
					if (this.matrix[i][j] > this.maxValue) {
						this.matrix[i][j] -= 4;
						if (i + 1 <= this.size - 1) {
							this.matrix[i+1][j] += 1;
						}
						if (i - 1 >= 0) {
							this.matrix[i-1][j] += 1;
						}
						if (j + 1 <= this.size - 1) {
							this.matrix[i][j+1] += 1;
						}
						if (j - 1 >= 0) {
							this.matrix[i][j-1] += 1;
						}
					}
				}
			}
		}
	}

	this.grow = function() {
		newMatrix = []
		for (i = 0; i < this.size + 2; i++) {
			row = []
			for (j = 0; j < this.size + 2; j++) {
				row.push(0);
			}
			newMatrix.push(row);
		}
		for (i = 0; i < this.size; i++) {
			for (j = 0; j < this.size; j++) {
				newMatrix[i+1][j+1] = this.matrix[i][j];
			}
		}
		this.matrix = newMatrix;
		this.size += 2;
		this.squareSize = 300/this.size;
	}

}

File: CC_01 Sandpile/Python/Sandpile.py
import Helper

class Sandpile:

	def __init__(self, matrix, size=3, maxValue=3):
		self.matrix = matrix
		self.size = size
		self.maxValue = maxValue

		self.show()
		while not self.stable():
			self.topple()
		print "Sandpile created!"


	def topple(self):
		for i in range(self.size):
			for j in range(self.size):
				if self.matrix[i][j] > self.maxValue:
					self.matrix[i][j] -= 4
					if i+1 <= self.size - 1:
						self.matrix[i+1][j] += 1
					if i-1 >= 0:
						self.matrix[i-1][j] += 1
					if j+1 <= self.size - 1:
						self.matrix[i][j+1] += 1
					if j-1 >= 0:
						self.matrix[i][j-1] += 1
					self.show()


	def show(self):
		for i in range(self.size):
			for j in range(self.size):
				print self.matrix[i][j],
			print
		print


	def stable(self):
		for i in range(self.size):
			for j in range(self.size):
				if self.matrix[i][j] > self.maxValue:
					return False
		return True


	def add(self, other):
		newMatrix = Helper.zeromatrix(self.size)
		for i in range(self.size):
			for j in range(self.size):
				newMatrix[i][j] =  self.matrix[i][j] + other.matrix[i][j]
		return Sandpile(newMatrix, self.size, self.maxValue)


	def visualise(self)
		

